relay-server/src/metrics_extraction/transactions/types.rs

```diff
@@ -1,20 +1,219 @@
+                timestamp,
+                tags.into(),
+            ),
+            TransactionMetric::CountPerRootProject { value, tags } => Metric::new(
+                MetricResourceIdentifier {
+                    ty: MetricType::Counter,
+                    namespace,
+                    name: "count_per_root_project",
+                    unit: MetricUnit::None,
+                },
+                MetricValue::Counter(value),
+                timestamp,
+                tags.into(),
+            ),
+            TransactionMetric::Duration { unit, value, tags } => Metric::new(
+                MetricResourceIdentifier {
+                    ty: MetricType::Distribution,
+                    namespace,
+                    name: "duration",
+                    unit: MetricUnit::Duration(unit),
+                },
+                MetricValue::Distribution(value),
+                timestamp,
+                tags.into(),
+            ),
+            TransactionMetric::Measurement {
+                name: kind,
+                value,
+                unit,
+                tags,
+            } => Metric::new(
+                MetricResourceIdentifier {
+                    ty: MetricType::Distribution,
+                    namespace,
+                    name: format!("measurements.{kind}").as_str(),
+                    unit,
+                },
+                MetricValue::Distribution(value),
+                timestamp,
+                tags.into(),
+            ),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+struct TransactionMeasurementTags {
+    measurement_rating: Option<String>,
+    universal_tags: BTreeMap<UniversalTags, String>,
+}
+
+impl From<TransactionMeasurementTags> for BTreeMap<String, String> {
+    fn from(value: TransactionMeasurementTags) -> Self {
+        let mut map = UniversalTags::into_str_keys(value.universal_tags);
+        if let Some(decision) = value.measurement_rating {
+            map.insert("measurement_rating".to_string(), decision);
+        }
+        map
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct TransactionBreakdownTags {
+    universal_tags: BTreeMap<UniversalTags, String>,
+}
+
+impl From<TransactionBreakdownTags> for BTreeMap<String, String> {
+    fn from(value: TransactionBreakdownTags) -> Self {
+        UniversalTags::into_str_keys(value.universal_tags)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct TransactionCPRTags {
+    decision: String,
+    universal_tags: BTreeMap<UniversalTags, String>,
+}
+
+impl From<TransactionCPRTags> for BTreeMap<String, String> {
+    fn from(value: TransactionCPRTags) -> Self {
+        let mut map = UniversalTags::into_str_keys(value.universal_tags);
+        map.insert("decision".to_string(), value.decision);
+        map
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct TransactionDurationTags {
+    universal_tags: BTreeMap<UniversalTags, String>,
+}
```

Comment:
nit: Define a single `UniversalTags` newtype used in both `user` and `duration` metrics. On that, you can implement `From<...> for BTreeMap` again.

`struct UniversalTags(BTreeMap<UniversalTag, String>)`